File: Leet 2904/solution.py
def shortest_beautiful_substring(s: str, k: int) -> str:
    counter: dict[str, int] = {"0": 0, "1": 0}
    n: int = len(s)

    result_len: int = -1
    result_str: str = ""

    p1: int = 0
    for p2 in range(n):
        counter[s[p2]] = counter.get(s[p2], 0) + 1

        while p1 < p2 and counter["1"] > k:
            counter[s[p1]] -= 1
            p1 += 1

        if counter["1"] == k:
            beautiful_str: str = s[p1 : p2 + 1].lstrip("0")
            beautiful_len: int = len(beautiful_str)
            print(beautiful_str)
            
            if result_len == -1:
                result_len = beautiful_len
                result_str = beautiful_str
            elif (beautiful_len, beautiful_str) < (result_len, result_str):
                result_len = beautiful_len
                result_str = beautiful_str

    return result_str

def main() -> None:
    print(shortest_beautiful_substring("100011001", 3))

if __name__ == "__main__":
    main()

File: Leet 2904/solution.ts
function shortestBeautifulSubstring(s: string, k: number): string {
    let counter = new Map<string, number>([
        ["0", 0],
        ["1", 0],
    ]);

    let resultStr: string = "";

    let p1: number = 0;
    for (let p2 = 0; p2 < s.length; p2++) {
        let p2Value = counter.get(s[p2]) ?? 0;
        counter.set(s[p2], p2Value + 1);

        while (counter.get("1")! > k) {
            let p1Value = counter.get(s[p1])!;
            counter.set(s[p1], p1Value - 1);
            p1++;
        }

        if (counter.get("1")! === k) {
            // strip leading zero's from beautiful.
            let beautifulStr: string = s.slice(p1, p2 + 1).replace(/^0+/, "");

            if (
                resultStr === "" ||
                beautifulStr.length < resultStr.length ||
                (beautifulStr.length === resultStr.length && beautifulStr < resultStr)
            ) {
                resultStr = beautifulStr;
            }
        }
    }

    return resultStr;
}

function main(): void {
    console.log("2904. Shortest and Lexicographically Smallest Beautiful String")

    console.log(shortestBeautifulSubstring("100011001", 3), "11001");
    console.log(shortestBeautifulSubstring("1011", 2), "11");
    console.log(shortestBeautifulSubstring("01011101000111110", 5), "11111");
}

main();